packages/engine/src/lib/helper/logging-utils.ts

```diff
@@ -1,123 +1,95 @@
 import sizeof from 'object-sizeof'
-import { isMemoryFilePath } from '../services/files.service'
 
 const TRUNCATION_TEXT_PLACEHOLDER = '(truncated)'
 const MAX_SINGLE_SIZE_FOR_SINGLE_ENTRY = 1024 * 1024
 
 export const loggingUtils = {
     async trimExecution(steps: Record<string, StepOutput>): Promise<Record<string, StepOutput>> {
         const clonedSteps = { ...steps }
-        for (const stepName in steps) {
-            const stepOutput = steps[stepName]
-            clonedSteps[stepName] = await trimStepOutput(stepOutput, 'trim')
-        }
-        // The above code could trim data in different steps, but their total size could still be too large.
-        return emptyStepsUntilSizeLimit(clonedSteps)
+        const jsonSteps = trimJson(JSON.parse(JSON.stringify(clonedSteps)))
+        const recordSteps: Record<string, StepOutput> = jsonSteps
+        return recordSteps
     },
 }
 
-async function trimStepOutput(stepOutput: StepOutput, mode: 'trim' | 'empty'): Promise<StepOutput> {
-    const modified: StepOutput = JSON.parse(JSON.stringify(stepOutput))
-    modified.input = await trimOrEmpty(mode, modified.input)
-    switch (modified.type) {
-        case ActionType.BRANCH:
-        case TriggerType.EMPTY:
-        case TriggerType.PIECE:
-        case ActionType.CODE:
-        case ActionType.PIECE:
-            modified.output = await trimOrEmpty(mode, modified.output)
-            break
-        case ActionType.LOOP_ON_ITEMS: {
-            const loopItem = modified.output
-            if (loopItem) {
-                loopItem.iterations = await trimIterations(loopItem.iterations, mode)
-                loopItem.item = await applyFunctionToValues(loopItem.item, trim)
-            }
-            break
-        }
-    }
-    modified.errorMessage = await trimOrEmpty(mode, modified.errorMessage)
-    return modified
-}
-
-async function trimOrEmpty(mode: 'trim' | 'empty', output: unknown): Promise<unknown> {
-    switch (mode) {
-        case 'trim':
-            return applyFunctionToValues(output, trim)
-        case 'empty':
-            return TRUNCATION_TEXT_PLACEHOLDER
+function trimJson(json: any){
+    type Node = {
+        size: number,
+        index: number,
+        parentNodeId: number,
+        numberOfChildren: number,
+        truncate: boolean
     }
-}
-
-async function trimIterations(iterations: Record<string, StepOutput>[], mode: 'trim' | 'empty'): Promise<Record<string, StepOutput>[]> {
-    const newIterations = []
-    for (const iteration of iterations) {
-        let trimmedIteration: Record<string, StepOutput> = {}
-        switch (mode) {
-            case 'empty':
-                for (const key in iteration) {
-                    trimmedIteration[key] = await trimStepOutput(iteration[key], mode)
-                }
-                break
-            case 'trim':
-                trimmedIteration = await loggingUtils.trimExecution(iteration)
-                break
+    
+    const nodes: Node[] = []
```

Comment:
When you making nodes global, you are making jsonToNode and other functions not pure as they depend on out nodes which life cycle is not controlled by them